tfjs-backend-webgpu/src/kernels/tile_webgpu.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+import {util} from '@tensorflow/tfjs-core';
+
+import {getCoordsDataType} from '../shader_preprocessor';
+import {computeDispatch, flatDispatchLayout} from '../webgpu_util';
+
+import {WebGPUProgram} from './webgpu_program';
+
+export class TileProgram implements WebGPUProgram {
+  variableNames = ['A'];
+  outputShape: number[];
+  userCode: string;
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  workGroupSize: [number, number, number] = [64, 1, 1];
+  dtype: string;
+  size: number;
+  rank: number;
+  aShape: number[];
+
+  constructor(aShape: number[], reps: number[]) {
+    const outputShape: number[] = new Array(aShape.length);
+    for (let i = 0; i < outputShape.length; i++) {
+      outputShape[i] = aShape[i] * reps[i];
+    }
+    this.outputShape = outputShape;
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workGroupSize);
+    this.aShape = aShape;
+    this.rank = this.outputShape.length;
+    this.size = util.sizeFromShape(this.outputShape);
+    this.shaderKey = 'tile';
+  }
+
+  getUserCode(): string {
+    const dtype = getCoordsDataType(this.rank);
+    const sourceCoords = getSourceCoords(this.aShape);
+
+    const userCode = `
+      void main() {
+        ${dtype} resRC = getOutputCoords();
+        int index = int(gl_GlobalInvocationID.x);
+        if (index < size) {
+          setOutput(index, getA(${sourceCoords}));
+        }
+      }
+    `;
+    return userCode;
+  }
+}
+
+function getSourceCoords(aShape: number[]): string {
```

Comment:
`aShape: number[]` is not needed. You can directly use `this.rank`? I think `aShape.length` equals `this.outputShape.length`?

Reply:
done